File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/abcpath/data/ABCPathDocument.java
package com.zwstudio.logicpuzzlesandroid.puzzles.abcpath.data;

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocument;
import com.zwstudio.logicpuzzlesandroid.common.data.MoveProgress;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.puzzles.abcpath.domain.ABCPathGame;
import com.zwstudio.logicpuzzlesandroid.puzzles.abcpath.domain.ABCPathGameMove;

import org.androidannotations.annotations.EBean;

@EBean
public class ABCPathDocument extends GameDocument<ABCPathGame, ABCPathGameMove> {
    protected void saveMove(ABCPathGameMove move, MoveProgress rec) {
        rec.row = move.p.row;
        rec.col = move.p.col;
        rec.strValue1 = String.valueOf(move.obj);
    }
    public ABCPathGameMove loadMove(MoveProgress rec) {
        return new ABCPathGameMove() {{
            p = new Position(rec.row, rec.col);
            obj = rec.strValue1.charAt(0);
        }};
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/abcpath/data/ABCPathDocument.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.abcpath.data

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocument
import com.zwstudio.logicpuzzlesandroid.common.data.MoveProgress
import com.zwstudio.logicpuzzlesandroid.common.domain.Position
import com.zwstudio.logicpuzzlesandroid.puzzles.abcpath.domain.ABCPathGame
import com.zwstudio.logicpuzzlesandroid.puzzles.abcpath.domain.ABCPathGameMove
import org.androidannotations.annotations.EBean

@EBean
class ABCPathDocument : GameDocument<ABCPathGame, ABCPathGameMove>() {
    override fun saveMove(move: ABCPathGameMove, rec: MoveProgress) {
        rec.row = move.p.row
        rec.col = move.p.col
        rec.strValue1 = move.obj.toString()
    }

    override fun loadMove(rec: MoveProgress) =
        ABCPathGameMove(Position(rec.row, rec.col), rec.strValue1!![0])
}
